google-cloud-core/src/main/java/com/google/cloud/Timestamp.java

```diff
@@ -137,22 +169,119 @@ public static Timestamp parseTimestamp(String timestamp) {
     }
     String secondsPart = matcher.group(1);
     String nanosPart = matcher.group(2);
-    long seconds;
-    seconds = format.parseMillis(secondsPart) / 1000;
+    long seconds = TimeUnit.MILLISECONDS.toSeconds(ISO_FORMAT_UTC.parseMillis(secondsPart));
     int nanos = 0;
     if (nanosPart != null) {
-      String padded = Strings.padEnd(nanosPart.substring(1), 9, '0');
-      nanos = Integer.parseInt(padded);
+      nanos = parseFractToNanos(nanosPart.substring(1));
+    }
+    return ofTimeSecondsAndNanos(seconds, nanos);
+  }
+
+  /**
+   * Creates a Timestamp instance from the given string. String is in the format
+   * yyyy-MM-dd HH:mm:ss.SSSSSSZZ. If no timezone is given, UTC is assumed.
+   */
+  public static Timestamp parseCloudTimestamp(String timestamp) {
+    String date = null;
+    String time = null;
+    String fract = null;
+    String tz = null;
+
+    Matcher dateMatcher = CLOUD_DATE_FORMAT_REGEXP.matcher(timestamp);
+    if (dateMatcher.matches()) {
+      date = dateMatcher.group(1);
+      tz = dateMatcher.group(2);
+    } else {
+      Matcher dateTimeMatcher = CLOUD_DATE_TIME_FORMAT_REGEXP.matcher(timestamp);
+      if (dateTimeMatcher.matches()) {
+        date = dateTimeMatcher.group(1);
+        time = dateTimeMatcher.group(2);
+        tz = dateTimeMatcher.group(3);
+      } else {
+        Matcher dateTimeFractMatcher = CLOUD_DATE_TIME_FRACT_FORMAT_REGEXP.matcher(timestamp);
+        if (dateTimeFractMatcher.matches()) {
+          date = dateTimeFractMatcher.group(1);
+          time = dateTimeFractMatcher.group(2);
+          fract = dateTimeFractMatcher.group(3);
+          tz = dateTimeFractMatcher.group(4);
+        } else {
+          throw new IllegalArgumentException("Cannot parse input: " + timestamp);
+        }
+      }
+    }
+
+    DateTime dateTime = new DateTime();
+    dateTime = parseTz(dateTime, tz);
+    dateTime = parseDate(dateTime, Preconditions.checkNotNull(date));
```

Comment:
That was a check that came along for the ride when I moved a bunch of code around. I will remove it.